File: c++/RecursiveChange.cpp
#include "stdafx.h"
#include "RecursiveChange.h"

using namespace std;

int dpChanges(int money, int coins[])
{
	int* minChanges = new int[money + 1];
	minChanges[0] = 0;
	int numCoin = 0;

	for (int i = 1; i < money + 1; i++) {
		minChanges[i] = 99999999;
		for (int j = 0; j < 3; j++) {
			if (i >= coins[j]) {
				numCoin = minChanges[i - coins[j]] + 1;
				if (numCoin < minChanges[i])
					minChanges[i] = numCoin;
			}
		}
	}

	return minChanges[money];
}

RecursiveChange::RecursiveChange()
{
	int money;
	cin >> money;
	
	int coins[] = { 4, 3, 1 };

	cout << dpChanges(money, coins) << endl;
}


RecursiveChange::~RecursiveChange()
{
}

File: java/src/RecursiveChange.java
import java.util.*;

public class RecursiveChange {
    private static int dpChanges(int money, int[] coins) {
        int[] minCoins = new int[money + 1];
        minCoins[0] = 0;
        int numCoin = 0;

        for (int i = 1 ; i < money + 1 ; i++) {
            minCoins[i] = 99999999;
            for (int coin : coins) {
                if (i >= coin) {
                    numCoin = minCoins[i - coin] + 1;
                    if (numCoin < minCoins[i])
                        minCoins[i] = numCoin;
                }
            }
        }

        return minCoins[money];
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int money = scanner.nextInt();
        int[] coins = {4,3,1};

        System.out.println(dpChanges(money, coins));
    }
}
